builder.md

```diff
@@ -1,57 +1,149 @@
+## Introduction
 
-## File Structure
+A base16 builder is, essentially, an application that can build base16
+templates with base16 [schemes](./styling.md).
 
-### Schemes Repository
+A template is a "blueprint" that specifies a file representing how to use the
+16 base16 colors with that specific software/format. For example: a
+[colorscheme template for vim](https://github.com/base16-project/base16-vim).
 
-The schemes repo should either be stored in a common location (perhaps referred to by environment variable or command line flag) or dynamically embedded in the builder.
+A scheme is a color palette that consists of 16 colors (hence the name). For
+example: [the solarized scheme](https://github.com/base16-project/base16-schemes/blob/main/solarized-dark.yaml)
 
-- `/*.yaml`
+Builders are designed for lower-level ("plumbing") usage, specifically for
+scripting and as component to build more complex base16 applications.
 
-### Template Repository
+The more complex apps designed for end users ("porcelain") are usually referred
+to as **managers**, and they don't need to follow any standard besides helping
+with base16 theming somehow, in a way that makes sense for their intended
+usecase.
 
-Each template repository should have a templates folder containing a config.yaml and any needed mustache template files.
+Template maintainers SHOULD provide built versions (with all existing scheme)
+so the end user doesn't need to be aware of the builder.
 
-- `/templates/*.mustache` - A template file (there may be multiples of these)
-- `/templates/config.yaml` - A template configuration file
+## Interface
 
-## Workflow
-The first thing a builder needs to do is parse all the scheme files from the schemes repository (as defined in the [file guidelines](https://github.com/chriskempson/base16/blob/master/file.md)). All files matching the pattern `*.yaml` should be loaded from the root of the schemes repository.
+All base16 builders MUST provide a single feature: building a template using 1 or more schemes.
 
-When building a target template repository, a base16 builder should first clear out any old output. Next, for all templates defined in the template repo's config file (located at `/templates/config.yaml`), the builder should iterate through all the defined schemes and output matching files. The built filename should look like `[output-dir]/base16-[slug][extension]`, where the slug is taken from the scheme filename made lowercase with spaces replaced with dashes and both the extension and output-dir are taken from `/template/config.yaml`.
+### CLI
 
-In the case where schemes share the same slug, a builder will overwrite files previously generated from the template. Should this happen, a builder should show warning messages listing the overwritten files.
+It is REQUIRED that that this functionality is exposed as binary CLI executable
```

Comment:
Does CLI need to be required for builder? I propose we just have one CLI implementation. Other builders just act as a library for different languages. And the CLI implementation can be thought of as a library for bash. And details of what the CLI looks like is a decision for the sole CLI implementation.

Reply:
Maybe we can require the builder to expose its functionality as CLI and/or Library.

I'm not sure if a single CLI implementation would be very good. What implementation should be the canonical one? And it's pretty easy to implement a CLI after you have the library, so perhaps it isn't too much of a requirement.

Reply:
Probably should move discussion to #13

Reply:
Yeah, I started that before I saw any of this.